python/cuml/fil/fil.pyx

```diff
@@ -0,0 +1,340 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+# cython: profile=False
+# distutils: language = c++
+# cython: embedsignature = True
+# cython: language_level = 3
+
+import copy
+import cudf
+import ctypes
+import math
+import numpy as np
+import warnings
+
+from numba import cuda
+
+from libcpp cimport bool
+from libc.stdint cimport uintptr_t
+from libc.stdlib cimport calloc, malloc, free
+
+from cuml.common.base import Base
+from cuml.common.handle cimport cumlHandle
+from cuml.utils import get_dev_array_ptr, input_to_dev_array, zeros
+cimport cuml.common.handle
+cimport cuml.common.cuda
+
+cdef extern from "treelite/c_api.h":
+    ctypedef void* ModelHandle
+    cdef int TreeliteLoadXGBoostModel(const char* filename,
+                                      ModelHandle* out)
+    cdef int TreeliteLoadXGBoostModelFromMemoryBuffer(const void* buf,
+                                                      size_t len,
+                                                      ModelHandle* out)
+    cdef int TreeliteFreeModel(ModelHandle handle)
+    cdef int TreeliteQueryNumTree(ModelHandle handle, size_t* out)
+    cdef int TreeliteQueryNumFeature(ModelHandle handle, size_t* out)
+    cdef int TreeliteLoadLightGBMModel(const char* filename, ModelHandle* out)
+    cdef int TreeliteLoadProtobufModel(const char* filename, ModelHandle* out)
+
+
+cdef class TreeliteModel():
+    """
+    Wrapper for Treelite-loaded forest
+
+    This is only used for loading saved models into FIL, it is
+    not used directly for inference.
+
+    Attributes
+    ----------
+
+    handle : ModelHandle
+        Opaque pointer to Treelite model
+    """
+    cpdef ModelHandle handle
+
+    def __cinit__(self):
+        self.handle = <ModelHandle>NULL
+
+    cdef set_handle(self, ModelHandle new_handle):
+        self.handle = new_handle
+
+    cdef ModelHandle get_handle(self):
+        return self.handle
+
+    def __cdel__(self):
+        if self.handle != NULL:
+            TreeliteFreeModel(self.handle)
+
+    @property
+    def num_trees(self):
+        assert self.handle != NULL
+        cdef size_t out
+        TreeliteQueryNumTree(self.handle, &out)
+        return out
+
+    @property
+    def num_features(self):
+        assert self.handle != NULL
+        cdef size_t out
+        TreeliteQueryNumFeature(self.handle, &out)
+        return out
+
+    @staticmethod
+    def from_filename(filename, model_type="xgboost"):
+        """
+        Returns a TreeliteModel object loaded from `filename`
+
+        Parameters
+        ----------
+        filename : string
+            Path to treelite model file to load
+
+        model_type : string
+            Type of model: 'xgboost', 'protobuf', or 'lightgbm'
+        """
+        filename_bytes = filename.encode("UTF-8")
+        cdef ModelHandle handle
+        if model_type == "xgboost":
+            res = TreeliteLoadXGBoostModel(filename_bytes, &handle)
```

Comment:
A wrapper that checks the return value and throws an exception would be better.

Reply:
This code does check the return value and raise an exception on failure.